src/main/java/com/hms_networks/sc/cumulocity/api/CConnectorProvisionMqttMgr.java

```diff
@@ -0,0 +1,195 @@
+      String mqttMessageDevicePassword = (String) mqttMessagePayloadSplit.get(3);
+
+      // Check if message is a device provisioning credentials message
+      if (mqttMessageId.equals(CUMULOCITY_MQTT_PROVISION_CREDENTIALS_MESSAGE_ID)) {
+        // Store new device credentials
+        try {
+          CConnectorMain.getConnectorConfig().setCumulocityDeviceTenant(mqttMessageDeviceTenant);
+          CConnectorMain.getConnectorConfig()
+              .setCumulocityDeviceUsername(mqttMessageDeviceUsername);
+          CConnectorMain.getConnectorConfig()
+              .setCumulocityDevicePassword(mqttMessageDevicePassword);
+          synchronized (this) {
+            isProvisioned = true;
+            notifyAll();
+          }
+        } catch (JSONException e) {
+          Logger.LOG_CRITICAL("Unable to save new device credentials during provisioning!");
+          Logger.LOG_EXCEPTION(e);
+        }
+      }
+    }
+  }
+
+  /** Method for processing or performing tasks on the looping MQTT thread. */
+  public void runOnMqttLoop() {
+    // Send empty message to Cumulocity (if subscribed to receive response)
+    if (subscribedToTopics) {
+      try {
+        mqttPublish(CUMULOCITY_MQTT_TOPIC_UCR, "", MQTT_QOS_LEVEL, true);
+      } catch (Exception e) {
+        Logger.LOG_CRITICAL("Unable to send device provisioning empty message to Cumulocity!");
+        Logger.LOG_EXCEPTION(e);
+      }
+    }
+  }
+
+  /**
+   * Method for managing MQTT status changes.
+   *
+   * @param status MQTT status code
+   */
+  public void onStatus(int status) {
+    Logger.LOG_CRITICAL("MQTT device provisioning client status changed to " + status);
+  }
+
+  /** Method for processing a successful MQTT connection. */
+  public void onConnect() {
+    Logger.LOG_CRITICAL("MQTT device provisioning client is connected!");
+
+    // Subscribe to Cumulocity MQTT topics
+    if (!subscribedToTopics) {
+      try {
+        subscribe(CUMULOCITY_MQTT_TOPIC_DCR, MQTT_QOS_LEVEL);
+        subscribedToTopics = true;
+      } catch (Exception e) {
+        Logger.LOG_CRITICAL(
+            "Unable to subscribe to Cumulocity MQTT topic: " + CUMULOCITY_MQTT_TOPIC_DCR);
+        Logger.LOG_EXCEPTION(e);
+      }
+    }
+  }
+
+  /**
+   * Method for waiting until device provisioning is complete. The calling thread is blocked until
+   * released on completion of the device provisioning process.
+   *
+   * @throws InterruptedException if the MQTT provisioning manager is interrupted
+   */
+  public synchronized void awaitProvisioning() throws InterruptedException {
+    if (!isProvisioned) {
+      wait();
+    }
+  }
+
+  /**
+   * Method for initializing connection
+   *
+   * @throws Exception if unable to initialize connection
+   */
+  public void initConnection() throws Exception {
+    Logger.LOG_DEBUG("Initializing MQTT device provisioning connection...");
+
+    // Configure MQTT
+    setPortNonSecureDefault();
```

Comment:
so this is done with TLS, but on port 1883?  Is there a 8883 option? This might be better received by IT personnel.

Reply:
The default MQTT HTTPS implementation for Cumulocity supports on two-way SSL with device certificates as well, and the documentation says one-way SSL with username/password for device authentication must be manually enabled or turned on by talking to Cumulocity support.